controllers/add_geteway.go

```diff
@@ -0,0 +1,179 @@
+	if hdb.Spec.Gateway == nil {
+		return nil
+	}
+
+	logger := log.WithValues("namespace", hdb.Namespace, "instance", hdb.Name, "reconciler", "add gateway")
+	if !hdb.Status.HServer.Bootstrapped {
+		return &requeue{message: "wait for HServer nodes to be ready", delay: 10 * time.Second}
+	}
+
+	deploy := a.getDeployment(ctx, r, hdb)
+
+	existingDeploy := &appsv1.Deployment{}
+	err := r.Client.Get(ctx, client.ObjectKeyFromObject(&deploy), existingDeploy)
+	if err != nil {
+		if !k8sErrors.IsNotFound(err) {
+			return &requeue{curError: err}
+		}
+		if err = ctrl.SetControllerReference(hdb, &deploy, r.Scheme); err != nil {
+			return &requeue{curError: err}
+		}
+
+		logger.Info("Create gateway")
+		if err = r.Client.Create(ctx, &deploy); err != nil {
+			return &requeue{curError: err}
+		}
+		return nil
+	}
+	if !isHashChanged(&existingDeploy.ObjectMeta, &deploy.ObjectMeta) {
+		return nil
+	}
+
+	logger.Info("Update gateway")
+	r.Recorder.Event(hdb, corev1.EventTypeNormal, "UpdatingGateway", "")
+
+	existingDeploy.Annotations = deploy.Annotations
+	existingDeploy.Labels = deploy.Labels
+	existingDeploy.Spec = deploy.Spec
+	if err = r.Update(ctx, existingDeploy); err != nil {
+		return &requeue{curError: err}
+	}
+	return nil
+}
+
+func (a addGateway) getDeployment(ctx context.Context, r *HStreamDBReconciler, hdb *hapi.HStreamDB) appsv1.Deployment {
+	podTemplate := a.getPodTemplate(ctx, r, hdb)
+	deploy := internal.GetDeployment(hdb, &hdb.Spec.Gateway.Component,
+		&podTemplate, hapi.ComponentTypeGateway)
+
+	return deploy
+}
+
+func (a addGateway) getPodTemplate(ctx context.Context, r *HStreamDBReconciler, hdb *hapi.HStreamDB) corev1.PodTemplateSpec {
+	gateway := hdb.Spec.Gateway
+
+	pod := corev1.PodTemplateSpec{
+		ObjectMeta: internal.GetObjectMetadata(hdb, nil, hapi.ComponentTypeGateway),
+		Spec: corev1.PodSpec{
+			Affinity:        gateway.Affinity,
+			Tolerations:     gateway.Tolerations,
+			NodeName:        gateway.NodeName,
+			NodeSelector:    gateway.NodeSelector,
+			SecurityContext: gateway.PodSecurityContext,
+			InitContainers:  gateway.InitContainers,
+			Containers:      a.getContainer(ctx, r, hdb),
+			Volumes:         gateway.Volumes,
+		},
+	}
+	if gateway.SecretRef != nil {
+		pod.Spec.Volumes = append(pod.Spec.Volumes, corev1.Volume{
+			Name: "cert",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: gateway.SecretRef.Name,
+				},
+			},
+		})
+	}
+
+	pod.Name = hapi.ComponentTypeAdminServer.GetResName(hdb.Name)
+	return pod
+}
+
+func (a addGateway) getContainer(ctx context.Context, r *HStreamDBReconciler, hdb *hapi.HStreamDB) []corev1.Container {
+	gateway := hdb.Spec.Gateway
+	container := corev1.Container{
+		Image:           hdb.Spec.Gateway.Image,
+		ImagePullPolicy: hdb.Spec.Gateway.ImagePullPolicy,
+	}
+
+	structAssign(&container, gateway.Container.DeepCopy())
```

Comment:
If doesn't use deep copy, will get panic `reflect: call of reflect.Value.Elem on struct Value`, I guess because the gateway is a pointer, not a struct ?